### PR TITLE
Fixing MonthCalendar accessibility tree

### LINKED ISSUE #4224: Incorrect MonthCalendar Accessibility tree at an increased size
* .NET Core Version: 6.0

* Have you experienced this same bug with .NET Framework?: Yes (works even worse)

**Problem description:**
- If a MonthCalendar has a big size, it has several months (years, centuries, depending on a mode) but Inspect see only a first part of the MonthCalendar (e.g. a first month). The tool doesn't see other parts of the MonthCalendar. Narrator can't navigate and announce these parts too.
image

**Expected behavior:**
- .NET MonthCalendar should have a correct accessibility tree, it should work like a native MonthCalendar: Accessibility tools (Narrator, NVDA, Inspect) can see and announce MonthCalendar items (buttons, days, months, years, etc.) when it is at an increased size.
image

**Minimal repro:**
- Create a .NET 6.0 app
- Add a MonthCalendar in the app
- Set a big size for the MonthCalendar (e.g. Size(400,400))
- Run the app and check its accessibility tree using Inspect

### PR DESCRIPTION
**Pull request by vladimir-krestov**
Fixes #4224

## Proposed changes

- Added a UI test app for testing a MonthCalendar control. 
- Remove all previous accessible objects implementations to create and set up MonthCalendar accessible objects from scratch. It is necessary for comfortable testing and review.
- Implement all MonthCalendar accessible objects with correct properties and actions. Add unit tests.
- Add a documentation described expected MonthCalendar accessible objects properties and actions depending on the calendar view. It is necessary for comfortable review, testing and implementing future fixes.

## Customer Impact

- Fixed incorrect accessibility properties of MonthCalendar items.
- MonthCalendar items are accessible in multi-calendar mode.

## Regression? 

- No

## Risk

- Minimal

### Before

- A MonthCalendar accessible object is correct only for one-month mode. 
image

### After

- All visible MonthCalendar items are accessible
image

## Test methodology 

- Manual testing
- Unit tests
- CTI

## Accessibility testing

- Using Inspect, Narrator and AccessibilityInsights

## Test environment(s) 

- .NET SDK Version: 6.0.100-preview.1.21101.5
- Microsoft Windows [Version 10.0.19042.867]

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/4758)

### CODE REVIEW

**Comment by vladimir-krestov** [comment]
Testers checked the fix. I have fixed all test issues. I need a code review before the final testing. 
@RussKie, @dreddy-work, could you please review the fix?

File: docs/accessibility/monthcalendar-expected-behavior.md [hunk 1]
```diff
@@ -0,0 +1,1359 @@
...
+
+
+<details>
+<summary>1. Month view</summary>
+</br>
+
+MonthCalendar:
+- [ ] `ControlType` = "calendar" always
+- [ ] `IsEnabled` = `true`, if the control is enabled
+- [ ] Has keyboard focus = `true`, if the control is in focus
```
Comment (RussKie) [inline docs/accessibility/monthcalendar-expected-behavior.md:80]: In other places you use `HasKeyboardFocus`, are they refer to the same thing? Is this a property like `ControlType`, etc.?

File: docs/accessibility/monthcalendar-expected-behavior.md [hunk 2]
```diff
@@ -0,0 +1,1359 @@
...
+- [ ] `IsEnabled` = `true`, if the control is enabled
+- [ ] Has default action and description
+- [ ] `Role` = "push button"
+- [ ] `State` = "normal"
+- [ ] Supports Invoke and LegacyIAccessible
+
+Calendar:
+- [ ] `IsEnabled` = `true`, if the control is enabled
+- [ ] Doesn't have a control type like as the native control
+- [ ] HasKeyboardFocus is true if the calendar contains the focused cell
```
Comment (RussKie) [inline docs/accessibility/monthcalendar-expected-behavior.md:117]: Should this be?
```suggestion
- [ ] `HasKeyboardFocus` = `true`, if the calendar contains the focused cell
```

File: docs/accessibility/monthcalendar-expected-behavior.md [hunk 3]
```diff
@@ -0,0 +1,1359 @@
...
+- [ ] `IsEnabled` = `true`, if the control is enabled
+- [ ] `ControlType` = "DataItem" ("item" in the accessibility tree)
+- [ ] `HasKeyboardFocus ` = `true`, if the cell is focused
+- [ ] `IsKeyboardFocusable` = `true`, if the control is enabled
+- [ ] Correct GridItem pattern properties
+- [ ] `Description` = "Week {number}, {day of week}" (e.g. "Week 10, Friday")
+- [ ] `Role` = "cell"
+- [ ] `State` = "focusable, selectable" if the control is enabled. (the order of the states doesn't matter)
+              "selected, focusable, selectable" if the cell is selected
+              "focused, selected, focusable, selectable" if the cell is selected and focused
```
Comment (RussKie) [inline docs/accessibility/monthcalendar-expected-behavior.md:189]: ℹ️ Having double space at the end of the line adds a soft paragraph break (i.e. `<br />`)
E.g.
image


Apply the same below, where necessary

File: docs/accessibility/monthcalendar-expected-behavior.md [hunk 4]
```diff
@@ -0,0 +1,1359 @@
...
+- [Behavior test cases](#Behavior-test-cases)
+- [Additional dev points that don't affect users](#Additional-dev-points-that-don't-affect-users)
+    
+
+# Overview
+
+[MonthCalendar](https://docs.microsoft.com/dotnet/api/system.windows.forms.monthcalendar) is a native Windows control, wrapped by Windows Forms SDK. Windows Forms provide a [UIA support](https://docs.microsoft.com/dotnet/framework/ui-automation/ui-automation-overview) for all
+MonthCalendar accessible objects and their children making the MonthCalendar control accessible for Windows Forms end users.
+
+Testing tools, and expected properties and actions in there are listed below.
```
Comment (RussKie) [inline docs/accessibility/monthcalendar-expected-behavior.md:27]: I'm struggling to parse the meaning of this sentence. Perhaps we need to reword it.

File: src/System.Windows.Forms.Primitives/src/Interop/Kernel32/Interop.SYSTEMTIME.cs [hunk 5]
```diff
@@ -16,6 +19,20 @@ public struct SYSTEMTIME
...
+                {
+                    Debug.Fail("Incorrect SYSTEMTIME info!");
+                    return DateTime.MinValue;
+                }
+
+                // DateTime gets DayOfWeek automatically
+                return new DateTime(sysTime.wYear,
+                    sysTime.wMonth, sysTime.wDay, sysTime.wHour,
+                    sysTime.wMinute, sysTime.wSecond, sysTime.wMilliseconds);
+            }
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/Kernel32/Interop.SYSTEMTIME.cs:35]: Tests?

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarAccessibleObject.cs [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            }
+
+            internal override UiaCore.IRawElementProviderFragment? FragmentNavigate(UiaCore.NavigateDirection direction)
+                => direction switch
+                {
+                    UiaCore.NavigateDirection.NextSibling
+                        => _monthCalendarAccessibleObject.CalendarsAccessibleObjects.Find(this)?.Next?.Value
+                        ?? (_monthCalendarAccessibleObject.ShowToday
+                        ? (AccessibleObject)_monthCalendarAccessibleObject.TodayLinkAccessibleObject
+                        : null),
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarAccessibleObject.cs:93]: ❗  Indent to correctly visualise the context

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarBodyAccessibleObject.cs [hunk 7]
```diff
@@ -10,66 +14,216 @@ namespace System.Windows.Forms
     public partial class MonthCalendar
     {
         /// <summary>
-        /// Represents the calendar body accessible object.
+        ///  Represents a calendar body accessible object of a MonthCalendar control.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarBodyAccessibleObject.cs:17]: What's a "calendar body"?
```suggestion
        ///  Represents an accessible object for a calendar body in <see cref="MonthCalendar"/> control.
```

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarBodyAccessibleObject.cs [hunk 8]
```diff
@@ -10,66 +14,216 @@ namespace System.Windows.Forms
...
-                    UiaCore.UIA.IsGridPatternAvailablePropertyId => true,
-                    UiaCore.UIA.IsTablePatternAvailablePropertyId => true,
+                    UiaCore.UIA.ControlTypePropertyId => UiaCore.UIA.TableControlTypeId,
+                    UiaCore.UIA.IsGridPatternAvailablePropertyId => IsPatternSupported(UiaCore.UIA.GridPatternId),
+                    UiaCore.UIA.IsTablePatternAvailablePropertyId => IsPatternSupported(UiaCore.UIA.TablePatternId),
                     _ => base.GetPropertyValue(propertyID)
                 };
+
+            /// <remark>
+            ///  A calendar has row headers (week numbers) if ShowWeekNumbers is true
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarBodyAccessibleObject.cs:139]:
```suggestion
            ///  A calendar has row headers (week numbers) if <see cref="ShowWeekNumbers"/> is true
```

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarButtonAccessibleObject.cs [hunk 9]
```diff
@@ -2,51 +2,135 @@
...
+using System.Runtime.InteropServices;
 using static Interop;
 
 namespace System.Windows.Forms
 {
     public partial class MonthCalendar
     {
-        internal abstract class CalendarButtonAccessibleObject : CalendarChildAccessibleObject
+        /// <summary>
+        ///  Represents a base button accessible object for buttons in a MonthCalendar control.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarButtonAccessibleObject.cs:14]:
```suggestion
        ///  Represents an accessible object for buttons in <see cref="MonthCalendar"/> control.
```

Ditto in other places

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarWeekNumberCellAccessibleObject.cs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        /// <summary>
+        ///  Represents a calendar week number cell accessible object of a MonthCalendar control.
+        /// </summary>
+        internal class CalendarWeekNumberCellAccessibleObject : CalendarCellAccessibleObject
+        {
+            // This const is used to get ChildId.
+            // It should take into account previous cells in a row.
+            // Indices start at 1.
+            private const int ChildId = 0;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarWeekNumberCellAccessibleObject.cs:19]: If "indices start at 1" why the value is 0?

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs [hunk 11]
```diff
@@ -214,613 +109,405 @@ internal override int ColumnCount
...
+                return hitTestInfo.uHit switch
+                {
+                    MCHT.CALENDARCONTROL => this,
+                    MCHT.TITLEBTNPREV or MCHT.PREV => PreviousButtonAccessibleObject,
+                    MCHT.TITLEBTNNEXT or MCHT.NEXT => NextButtonAccessibleObject,
+                    MCHT.CALENDARDATEMIN // The given point was over the minimum date in the calendar
+                        => CalendarsAccessibleObjects.First?.Value is CalendarAccessibleObject firstCalendar
+                        && firstCalendar.Bounds.Contains(innerX, innerY)
+                        ? firstCalendar
+                        : this,
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs:135]: Incorrect indentation
```suggestion
                        => CalendarsAccessibleObjects.First?.Value is CalendarAccessibleObject firstCalendar
                        && firstCalendar.Bounds.Contains(innerX, innerY)
                            ? firstCalendar
                            : this,
```

Ditto below

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs [hunk 12]
```diff
@@ -214,613 +109,405 @@ internal override int ColumnCount
...
-            public unsafe MCHITTESTINFO GetHitTestInfo(int xScreen, int yScreen)
-            {
-                if (!_ownerMonthCalendar.IsHandleCreated || HandleInternal == IntPtr.Zero)
-                {
-                    return new MCHITTESTINFO();
-                }
+            internal CalendarCellAccessibleObject? FocusedCell
+                => UiaCore.UiaClientsAreListening().IsTrue()
+                ? _focusedCellAccessibleObject ??= GetCellByDate(_owningMonthCalendar._focusedDate)
+                : null;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs:165]: ditto

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs [hunk 13]
```diff
@@ -214,613 +109,405 @@ internal override int ColumnCount
...
+                    {
+                        continue;
+                    }
 
-                if (!success)
-                {
-                    calendarPartRectangle = new RECT();
+                    DateTime calendarStart = calendar.DateRange.Start;
+                    DateTime calendarEnd = calendar.DateRange.End;
+                    if (date >= calendarStart && date <= calendarEnd)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs:286]: Invert

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/MonthCalendar.CalendarBodyAccessibleObjectTests.cs [hunk 14]
```diff
@@ -2,29 +2,215 @@
...
-            Assert.Equal(calendarAccessibleObject, bodyAccessibleObject.Parent);
+            using MonthCalendar control = new MonthCalendar();
+            int calendarIndex = 0;
+
+            MonthCalendarAccessibleObject controlAccessibleObject = (MonthCalendarAccessibleObject)control.AccessibilityObject;
+            CalendarAccessibleObject calendarAccessibleObject = new(controlAccessibleObject, calendarIndex, "Test name");
+            CalendarBodyAccessibleObject accessibleObject = new(calendarAccessibleObject, controlAccessibleObject, calendarIndex);
+
+            Assert.Equal(calendarAccessibleObject, accessibleObject.Parent);
+        }
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/MonthCalendar.CalendarBodyAccessibleObjectTests.cs:27]: Here and other places - assert handle not created?

**Review by RussKie — COMMENTED** [review]
🚀

File: docs/accessibility/monthcalendar-expected-behavior.md (hunk 1 above)
Comment (vladimir-krestov) [reply docs/accessibility/monthcalendar-expected-behavior.md]: Yes, this is a property:

image

File: docs/accessibility/monthcalendar-expected-behavior.md (hunk 4 above)
Comment (vladimir-krestov) [reply docs/accessibility/monthcalendar-expected-behavior.md:27]: Reworded it like this:
"Expected properties and actions of a MonthCalendar and its parts in different accessibility tools are listed below."

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarWeekNumberCellAccessibleObject.cs (hunk 10 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.CalendarWeekNumberCellAccessibleObject.cs]: Added the following comment:
> "It should take into account next cells in a row.
Indices of date cells start at 1, so use 0 to note that the cell is a header."

**Comment by dreddy-work** [comment]
@Tanya-Solyanik , when you get time, can you please take a look at this?

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
I've had an offline chat with @merriemcgaw and agreed to have it merged asap to have a longer runway to iron out any outstanding kinks. The change is tested by the CTI folks, and was found to be of generally acceptable quality. Any outstanding issues will be fixed in BAU mode.

**Comment by dreddy-work** [comment]
Because branch was created before this. We need to dual check-in this.

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.cs [hunk 15]
```diff
@@ -1461,9 +1488,15 @@ protected override void OnGotFocus(EventArgs e)
         {
             base.OnGotFocus(e);
 
-            AccessibilityObject.RaiseAutomationEvent(UiaCore.UIA.AutomationFocusChangedEventId);
+            ((MonthCalendarAccessibleObject)AccessibilityObject).FocusedCell?.RaiseAutomationEvent(UiaCore.UIA.AutomationFocusChangedEventId);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.cs:1491]: @vladimir-krestov This forces creation of AccessibleObject for scenarios where this object is not needed, doesn't it? Please fix even though this is not a recent regression.

Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.cs:1491]: @Tanya-Solyanik  I made the following:
Control.cs:
image
MonthCalendar.cs:
image

It leads to breaking Narrator behaviour:

KsJM6KhKDy

|
|
|

**UiaCore.UiaClientsAreListening().IsTrue()** won't work here.

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.cs:1491]: Are you saying that narrator does not send WM_GETOBJECT?

Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.cs:1491]: It sends WM_GETOBJECT after a MonthCalendar gets focus. So the first time Narrator doesn't focuses on a calendar. If to focus on the first form and return focus on the MonthCalendar form, Narrator will work well

Comment (merriemcgaw) [reply src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.cs:1491]: Can we file an issue on this to address in the next Preview?

Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.cs:1491]: https://github.com/dotnet/winforms/issues/5018

**Pull request merged**